### PR TITLE
Enable compile-time checks (warnings)

### LINKED ISSUE #51: Compiler Warnings
We should start using compiler warnings.

Topics:

- What set of warnings do we want to use?
- Manage different compilers options
- Should warning be set just for internal build (PRIVATE) or also as a usage constraint (PUBLIC)?

### PR DESCRIPTION
**Pull request by albestro**
Close #51 

This PR aims to enable compiler time check for the library, following the discussion in the issue.

### CODE REVIEW

File: cmake/DLAF_AddTargetWarnings.cmake [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      -Wextra
+      -Wnon-virtual-dtor
+      -Wunused
+      -Woverloaded-virtual
+      -Wconversion
+      -pedantic-errors
+
+      # googletest macro problem
+      # must specify at least one argument for '...' parameter of variadic macro
+      -Wno-gnu-zero-variadic-macro-arguments
```
[the remark this answers is not in the document]
Comment (albestro) [reply cmake/DLAF_AddTargetWarnings.cmake]: About different compilers I'm going to use CMake generator expressions on the variable [`CMAKE_CXX_COMPILER_ID`](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_ID.html#variable:CMAKE_%3CLANG%3E_COMPILER_ID).

The question is: what compilers do we want to support?
- gcc
- clang
- intel?

I can work on a solution where not supported compilers will not add any compile-time check (e.g. MSVC).

[the remark this answers is not in the document]
Comment (albestro) [reply cmake/DLAF_AddTargetWarnings.cmake]: > How do you plan to deal with external libraries warnings?

See the very last section of https://github.com/eth-cscs/DLA-Future/issues/51#issuecomment-595288891 😉

**Comment by albestro** [comment]
About compilers, with @rasolca we've decided to support:
- gcc
- clang/apple-clang

**Comment by albestro** [comment]
There is still an open problem about the list of warnings to enable.

https://godbolt.org/z/8Zoytc

`-Wconversion`, despite the same name on both clang and gcc, produces a different set of warnings. In particular, gcc is way more strict than clang.

What do we want to do? Keep it and fix all of them? (in case we will think how later)

**Comment by rasolca** [comment]
`-Wconversion` on gcc is too restrictive (e.g. int to float conversion, which in general shouldn't be a problem). Unfortunately there is no way to only disable this warning, but integer to smaller integer warnings has to be disabled as well.
Suggested: `-Wsign-conversion -Wfloat-conversion`

On the other hand clang 9 never reports the float to integer warning (tested with -Weverything).

**Comment by rasolca** [comment]
Just noticed: clang11 seems to report int to float conversions as warning with `-Wconversion` (which enables the new warning `-Wimplicit-int-float-conversion`).
This has to be taken in account.

File: include/dlaf/common/vector.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  vector(int size) : std::vector<T>(to_sizet(size)) {}
+
+  T& operator[](int index) {
+    return std::vector<T>::operator[](to_sizet(index));
+  }
+
+  const T& operator[](int index) const {
+    return std::vector<T>::operator[](to_sizet(index));
+  }
+};
```
Comment (rasolca) [inline include/dlaf/common/vector.h]: I would keep it `internal`, to avoid use in public interfaces.

**Pull request merged**